### PR TITLE
remote: RFC

### PR DESCRIPTION
**Pull request by efiop**
Making this method separate from exists() makes it easier to optimize
requests for large amounts of cache, since we can re-use the same prefix
for retrieving all the necessary information in a one single request.

Related #1430

Signed-off-by: Ruslan Kuprieiev <ruslan@iterative.ai>

### CODE REVIEW

File: dvc/remote/azure.py [hunk 1]
```diff
@@ -246,8 +256,8 @@ def _path_to_etag(self, path):
         return posixpath.dirname(relpath) + posixpath.basename(relpath)
 
     def _all(self):
-        keys = self._all_keys(self.bucket, self.prefix)
-        return [self._path_to_etag(key) for key in keys]
+        for key in self._list_keys(self.bucket, self.prefix):
+            yield self._path_to_etag(key)
```
[the remark this answers is not in the document]
Comment (efiop) [reply dvc/remote/azure.py]: The list might be way too big(e.g. 100M entries, md5 for each is 32 bytes, so ~3200Mb list) and we don't really need all of it at the same time, so it makes sense to use a generator to gradually iterate over it, without keeping all of it in memory.

Great point! I'll replace this for loop with a list comprehension :+1:

File: dvc/remote/gs.py [hunk 2]
```diff
@@ -179,26 +175,36 @@ def md5s_to_path_infos(self, md5s):
...
+
+        return False
+
+    def cache_exists(self, md5s):
+        assert isinstance(md5s, list)
+
+        if len(md5s) == 0:
+            return []
+
+        ret = len(md5s) * [False]
```
[the remark this answers is not in the document]
Comment (efiop) [reply dvc/remote/gs.py]: Great point!

> return [key in cache for key in keys]

The problem here is that the order will not match the order in which keys are located in the `cache`. Though we can use a dict instead. Let me take a look...

File: dvc/remote/azure.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mroutis) [reply dvc/remote/azure.py]: Could you please add this as a comment, @efiop ? Just like a future reminder that it we may be dealing with a lot of files; sometimes I forgot that when reading the code :see_no_evil:

File: dvc/remote/gs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mroutis) [reply dvc/remote/gs.py]: You could also use `zip` to pair arrays:
```python
>>> list(zip([1,2], [3, 4]))
[(1, 3), (2, 4)]
```

File: dvc/remote/azure.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (efiop) [reply dvc/remote/azure.py]: Sure. Thanks for pointing that out! :slightly_smiling_face:

**Pull request merged**